### DISCUSSION TITLE
mocker fixture usage - assistance please

### DISCUSSION THREAD

**Discussion by kevindawson** [pytest-dev/pytest#7789] [Q&A] [answered]
Trying to come to terms with **mocking** AND pytest
the following is ok,

test code from 2020 youtube -> @nikhilkumarsingh  [tut9](https://github.com/nikhilkumarsingh/pytest-tut/tree/master/tut9)
```python
from unittest import mock
import pytest
from tut9.myapp.sample import guess_number


@pytest.mark.parametrize("_input,expected", [(3, "You won!"), (4, "You lost!")])
@mock.patch("tut9.myapp.sample.roll_dice")
def test_guess_number(mock_roll_dice, _input, expected):
    mock_roll_dice.return_value = 3
    assert guess_number(_input) == expected
    mock_roll_dice.assert_called_once()
```
but when I try to code it to use **mocker**

```python
import pytest
from tut9.myapp.sample import guess_number


@pytest.mark.parametrize("in_put,out_put", [(3, "You won!"), (4, "You lost!")])
def test_guess_num(mocker, in_put, out_put):
    mocker.patch("tut9.myapp.sample.roll_dice", return_value=3)
    assert out_put == guess_number(in_put)
    #tut9.myapp.sample.roll_dice.assert_called_once()
```
the above partial conversion has taken me hours
Why
the lack of/no examples with plan English text

i have been using the IDE intellisense

I have no idea as to how to implement ```.assert_called_once()```

all I can see are redirects to ```unittest.mock``` API which is of NO help at all as there examples do not translate

I might be old and senile now

what about the _teen pythonites_

**Comment by kevindawson** [comment]
https://docs.pytest.org/en/stable/search.html?q=mocker

```html
<div id="search-results">
<h2>Search Results</h2>
<p class="search-summary">Your search did not match any documents. Please make sure that all words are spelled correctly and that you've selected enough categories.</p>
<ul class="search"></ul>
</div>
````

**Comment by nicoddemus** [comment]
Hi @kevindawson,

`mocker` is a fixture from the [pytest-mock](https://github.com/pytest-dev/pytest-mock) plugin, and not a part of pytest, which explains why you didn't find any results in the pytest's documentation.

The `mocker` fixture is a thing wrapper around the mocker functionality, this is your original code using `mocker`:

```python
import pytest
from tut9.myapp.sample import guess_number


@pytest.mark.parametrize("_input,expected", [(3, "You won!"), (4, "You lost!")])
def test_guess_number(mocker, _input, expected):
    mock_roll_dice = mock.patch("tut9.myapp.sample.roll_dice")
    mock_roll_dice.return_value = 3
    assert guess_number(_input) == expected
    mock_roll_dice.assert_called_once()
```

Particularly I find using the decorator facilities of `unittest.mock` a bit confusing when used with pytest, that's why `pytest-mock` exists, see the [rationale here](https://github.com/pytest-dev/pytest-mock#why-bother-with-a-plugin).

**Comment by kevindawson** [comment]
@nicoddemus 
did you mean
```python
mock_roll_dice = mock.patch("tut9.myapp.sample.roll_dice")
mock_roll_dice.return_value=3
```
OR
```python
mock_roll_dice = mocker.patch("tut9.myapp.sample.roll_dice", return_value=3)
```

Also in your [Why bother with a plugin?](https://github.com/pytest-dev/pytest-mock#why-bother-with-a-plugin) you seam to be saying what not to do, rather than you could do this with an example

**Comment by nicoddemus** [comment]
> did you mean
> OR
> ...

My bad. Both will work though (updated the original post).

> Also in your Why bother with a plugin? you seam to be saying what not to do rather that you could do this with an example

There I try to exemplify how one would use `unittest.mock` directly and why I believe it is not a great fit with pytest, hence the plugin.

**Comment by kevindawson** [comment]
@nicoddemus 
1, are you sure about that update?
2, I know python ```unittest.mock``` is not a great **FULL STOP**
3, I could not find any examples exemplifying good practice - hence over to you ;)

4, I found some documentation: [Data Pipelines with Apache Airflow](https://livebook.manning.com/book/data-pipelines-with-apache-airflow/chapter-8/v-5/63) 8.1.4   pytest project structure. Cool
5, **discussions** will be more popular in time than #pylib is currently

**Comment by nicoddemus** [comment]
> 1, are you sure about that update?

Why do you ask? If you see something strange, please point it out, no need to make me review it entirely.

> 3, I could not find any examples exemplifying good practice - hence over to you ;)

Hence what over to me?

**Comment by kevindawson** [comment]
1, I think you should have 

Assign mock to a variable, enables capture behaviour
```python
mock_roll_dice = mock.patch("tut9.myapp.sample.roll_dice")
```
2, @nicoddemus you are the author are you not. you have written a cool **plugin** because you are a cool dude, but you have forgotten to turn the light on for all the new users, especially the teen pythonites

3, I will ask @nikhilkumarsingh about using his code so you can have some examples for us mere mortals when we first come across something NEW

**Comment by nicoddemus** [comment]
> Assign mock to a variable, enables capture behaviour

Ahh thanks. Fixed.

@kevindawson can you mark my first post as answering your question if that's the case? This will help others find the answer quickly. Thanks. 👍

**Comment by kevindawson** [comment, marked as answer]
@nicoddemus I have pondered over this for some time now:

1, the use of ```_input``` to avoid stomping ove the python built-in function(input)
* I am lead to believe -> it should have been ```input_``` both of which I do not like
* as mocking is akin to black boxes -> in_put and conversely out_put make more sense

2, Assign mock to a variable, enables capture behaviour
* but why O why do we have to prefix it with ```mock_roll_dice``` as per ```unittest```
* as ```pytest-mock``` let's us define the variable name we could be more descriptive ```loaded_dice``` in this case

3, the use of the helper method ```assert_called_once()```
* issues as per [engineroom](http://engineroom.trackmaven.com/blog/mocking-mistakes/) & [Threat or Menace](https://engineeringblog.yelp.com/2015/02/assert_called_once-threat-or-menace.html)
* plus you can observe the *variable* ```call_count``` and see it incrementing via the python debugger (pyCharm community ed.)

All of the above goes into the following:


```python
@pytest.mark.parametrize("in_put,out_put", [(3, "You won!"), (4, "You lost!")])
def test_guess_number(mocker, in_put, out_put):
    loaded_dice = mocker.patch("tut9.myapp.sample.roll_dice", return_value=3)
    assert guess_number(in_put) == out_put
    assert loaded_dice.call_count == 1
```
